packages/flutter/lib/src/rendering/paragraph.dart

```diff
@@ -110,8 +113,20 @@ class RenderParagraph extends RenderBox {
     markNeedsLayout();
   }
 
+  /// An optional maximum number of lines for the text to span, wrapping if necessary.
+  /// If the text exceeds the given number of lines, it will be truncated according
+  /// to [overflow].
+  int get maxLines => _maxLines;
+  int _maxLines;
+  set maxLines(int value) {
+    if (_maxLines == value)
+      return;
+    _maxLines = value;
+    markNeedsLayout();
+  }
+
   void _layoutText({ double minWidth: 0.0, double maxWidth: double.INFINITY }) {
-    bool wrap = _softWrap || _overflow == TextOverflow.ellipsis;
+    bool wrap = _softWrap || _overflow == TextOverflow.ellipsis; // @@@MP suspect
```

Comment:
I found the problem with this line.

If I have `Text(longLine, softWrap: false, maxLines: 3, overflow: TextOverflow.clip)`, then it will remain one 1 line (unless there are explicit line breaks). But if I have `Text(longLine, softWrap: false, maxLines: 3, overflow: TextOverflow.ellipsis)` then it will wrap to exactly 3 lines and truncate with an ellipsis there.

I've changed it to only force wrap if maxLines is null. This is weird and subtle, but I think is the best compromise. It allows you to say both "I don't care how many lines this takes up, but end it with an ellipsis if it exceeds the width" [1] and also "I want this to not exceed N lines, and end with an ellipsis if it exceeds N lines." [2]

[1] softWrap = false, maxLines = null
[2] softWrap = true, maxLines = N